File: api/middleware/access_control.py
"""
Access control middleware for LocalBooru network security.

Controls access based on client IP address:
- localhost: Full access (read + write)
- local_network: Full access (read + write) - same WiFi/LAN
- public: Read-only (public internet IPs)

Settings/admin endpoints are always localhost-only.

Authentication enforcement based on auth_required_level setting:
- none: No auth required for any access level
- public: Auth required only for public internet access
- local_network: Auth required for local_network AND public access
- always: Auth required for ALL access including localhost
"""
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import JSONResponse
from typing import Callable, Optional

from ..services.network import classify_ip
from ..services.auth import decode_token
from ..routers.settings import get_network_settings


# Endpoints that are always localhost-only (sensitive settings)
LOCALHOST_ONLY_PREFIXES = [
    "/api/settings",
    "/api/network",
    "/api/users",
]

# Read-only endpoints safe for remote access
READ_ONLY_ENDPOINTS = [
    "GET",  # All GET requests are read-only
]

# Endpoints that require write access
WRITE_METHODS = ["POST", "PUT", "PATCH", "DELETE"]

# Endpoints exempt from access control (health checks, static files)
EXEMPT_PREFIXES = [
    "/health",
    "/docs",
    "/openapi.json",
    "/redoc",
    "/assets",  # Static frontend assets
    "/icon.png",
]

# Endpoints exempt from authentication (chicken-egg problem for login)
AUTH_EXEMPT_ENDPOINTS = [
    "/api/users/login",
    "/api/users/verify",
    "/api/network/verify-handshake",
]

# Endpoints under localhost-only prefixes that should still be accessible from network
LOCALHOST_EXEMPTIONS = [
    "/api/network/verify-handshake",
]

# Auth level hierarchy - which access levels require auth at each setting
# Key: auth_required_level setting value
# Value: set of access levels that require authentication
AUTH_LEVEL_REQUIREMENTS = {
    "none": set(),  # No auth required for anyone
    "public": {"public"},  # Only public internet needs auth
    "local_network": {"local_network", "public"},  # LAN and public need auth
    "always": {"localhost", "local_network", "public"},  # Everyone needs auth
}


def get_client_ip(request: Request) -> str:
    """
    Extract the client IP from the request.

    Security: We intentionally ignore X-Forwarded-For and X-Real-IP headers
    because this app is designed for direct access, not behind a reverse proxy.
    Trusting these headers would allow attackers to spoof their IP as localhost.
    """
    # Use only the direct connection IP - never trust proxy headers
    if request.client:
        return request.client.host

    return "unknown"  # Don't default to localhost - that would grant access


def cors_response(status_code: int, content: dict) -> JSONResponse:
    """Create a JSON response with CORS headers for error responses."""
    return JSONResponse(
        status_code=status_code,
        content=content,
        headers={
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Methods": "*",
            "Access-Control-Allow-Headers": "*",
        }
    )


def extract_bearer_token(request: Request) -> Optional[str]:
    """
    Extract the Bearer token from the Authorization header.

    Returns the token string if present and properly formatted, None otherwise.
    """
    auth_header = request.headers.get("Authorization")
    if not auth_header:
        return None

    # Expected format: "Bearer <token>"
    parts = auth_header.split(" ", 1)
    if len(parts) != 2 or parts[0].lower() != "bearer":
        return None

    return parts[1]


def is_auth_exempt(path: str) -> bool:
    """Check if the endpoint is exempt from authentication requirements."""
    # Check exact matches for auth-exempt endpoints
    for exempt_path in AUTH_EXEMPT_ENDPOINTS:
        if path == exempt_path or path.startswith(exempt_path + "/"):
            return True
    return False


class AccessControlMiddleware(BaseHTTPMiddleware):
    """
    Middleware to enforce network access control based on client IP.

    Sets request.state.access_level to one of: 'localhost', 'local_network', 'public'
    Sets request.state.client_ip to the client's IP address

    Blocks:
    - Non-localhost access to sensitive endpoints (/settings, /network, /users)
    - Write operations from non-localhost unless explicitly allowed
    - Access levels that aren't enabled in settings
    """

    async def dispatch(self, request: Request, call_next: Callable):
        path = request.url.path
        method = request.method

        # Log ALL incoming requests for debugging
        client = request.client.host if request.client else "unknown"
        print(f"[AccessControl] INCOMING: {method} {path} from {client}")

        try:
            # Always allow OPTIONS requests (CORS preflight)
            if method == "OPTIONS":
                return await call_next(request)

            # Skip access control for exempt endpoints
            for prefix in EXEMPT_PREFIXES:
                if path.startswith(prefix):
                    return await call_next(request)

            # Get client IP and classify
            client_ip = get_client_ip(request)
            access_level = classify_ip(client_ip)
        except Exception as e:
            print(f"[AccessControl] EXCEPTION in middleware: {e}")
            import traceback
            traceback.print_exc()
            raise

        # Store in request state for use by route handlers
        request.state.client_ip = client_ip
        request.state.access_level = access_level
        request.state.user = None  # Will be set if authenticated

        # Debug logging for non-localhost access
        if access_level != "localhost":
            print(f"[AccessControl] {method} {path} from {client_ip} ({access_level})")

        # Get network settings (needed for both auth and access level checks)
        network_settings = get_network_settings()

        # Check authentication requirements based on auth_required_level setting
        auth_required_level = network_settings.get("auth_required_level", "none")
        access_levels_requiring_auth = AUTH_LEVEL_REQUIREMENTS.get(auth_required_level, set())

        # Determine if authentication is required for this request
        auth_required = access_level in access_levels_requiring_auth

        # Check for auth-exempt endpoints (login, verify, etc.)
        if auth_required and is_auth_exempt(path):
            auth_required = False
            print(f"[AccessControl] Auth exempt endpoint: {path}")

        # Process authentication if required or if token is provided
        token = extract_bearer_token(request)
        if token:
            user_payload = decode_token(token)
            if user_payload:
                request.state.user = user_payload
                print(f"[AccessControl] Authenticated user: {user_payload.get('username')}")
            elif auth_required:
                # Token provided but invalid/expired
                return cors_response(
                    401,
                    {
                        "error": "Invalid or expired token",
                        "detail": "The provided authentication token is invalid or has expired. Please log in again."
                    }
                )

        # If auth is required but no valid token, reject the request
        if auth_required and request.state.user is None:
            print(f"[AccessControl] Auth required but no valid token for {method} {path} from {access_level}")
            return cors_response(
                401,
                {
                    "error": "Authentication required",
                    "detail": f"This endpoint requires authentication for {access_level} access. Please provide a valid Bearer token in the Authorization header."
                }
            )

        # Localhost always has full access (after auth check)
        if access_level == "localhost":
            return await call_next(request)

        print(f"[AccessControl] Network settings: local_network_enabled={network_settings.get('local_network_enabled')}")

        # Check if this access level is enabled
        if access_level == "local_network":
            if not network_settings.get("local_network_enabled", False):
                return cors_response(
                    403,
                    {
                        "error": "Local network access is disabled",
                        "detail": "Enable local network access in settings to connect from LAN"
                    }
                )
        elif access_level == "public":
            if not network_settings.get("public_network_enabled", False):
                return cors_response(
                    403,
                    {
                        "error": "Public network access is disabled",
                        "detail": "Enable public network access in settings to connect from internet"
                    }
                )

        # Block localhost-only endpoints for non-localhost (with exemptions)
        for prefix in LOCALHOST_ONLY_PREFIXES:
            if path.startswith(prefix):
                # Check if this specific path is exempted from localhost restriction
                is_exempted = any(path == exempt or path.startswith(exempt + "/")
                                  for exempt in LOCALHOST_EXEMPTIONS)
                if not is_exempted:
                    return cors_response(
                        403,
                        {
                            "error": "This endpoint is only accessible from localhost",
                            "detail": f"Access to {prefix} requires direct access to the machine running LocalBooru"
                        }
                    )

        # Block write operations for public internet (local network gets full access)
        if method in WRITE_METHODS and access_level == "public":
            return cors_response(
                403,
                {
                    "error": "Write operations require local access",
                    "detail": "Public internet access is read-only. Modifications must be made from localhost or local network."
                }
            )

        # Allow the request
        return await call_next(request)


def check_localhost_only(request: Request) -> bool:
    """
    Dependency to ensure request is from localhost.

    Usage in routes:
        @router.post("/sensitive-action")
        async def sensitive_action(
            _: bool = Depends(check_localhost_only)
        ):
            ...
    """
    access_level = getattr(request.state, "access_level", None)
    if access_level != "localhost":
        from fastapi import HTTPException
        raise HTTPException(
            status_code=403,
            detail="This action is only available from localhost"
        )
    return True

File: src-tauri/src/server/middleware/access_control.rs
use axum::{
    body::Body,
    extract::ConnectInfo,
    http::{Request, Response, StatusCode},
    response::IntoResponse,
};
use std::future::Future;
use std::net::SocketAddr;
use std::path::{Path, PathBuf};
use std::pin::Pin;
use std::task::{Context, Poll};
use tower::{Layer, Service};

use super::auth::{decode_jwt, refresh_claim_identity, Claims};
use crate::db::pool::DbPool;

fn claim_allows_source(claims: &Claims, source: &str) -> bool {
    match claims.access_level.as_str() {
        "public" => true,
        "local_network" => source != "public",
        "localhost" => source == "localhost",
        _ => false,
    }
}

/// Endpoints that are always localhost-only (sensitive settings).
const LOCALHOST_ONLY_PREFIXES: &[&str] = &[
    "/api/settings",
    "/api/network",
    "/api/users",
    "/api/direct-files",
    "/api/device-pairing/desktop-sessions",
    "/api/device-pairing/devices",
];

/// Endpoints exempt from access control (prefix match).
const EXEMPT_PREFIXES: &[&str] = &[
    "/health",
    "/docs",
    "/assets",
    "/thumbnails",
    "/icon.png",
    "/api/share/",
    "/api/cast-media/",
    "/watch/",
];

/// Endpoints exempt from access control (exact match).
const EXEMPT_EXACT: &[&str] = &["/api"];

/// Endpoints under localhost-only prefixes that should still be accessible from network.
const LOCALHOST_EXEMPTIONS: &[&str] = &[
    "/api/network/verify-handshake",
    "/api/settings/saved-searches",
    "/api/settings/family-mode",
    "/api/settings/video-playback",
    "/api/settings/optical-flow/stop",
    "/api/settings/svp",
    "/api/settings/whisper",
    "/api/settings/cast",
    "/api/settings/transcode",
    "/api/settings/video-info",
    "/api/settings/util",
    "/api/users/login",
    "/api/users/verify",
    "/api/users/media-token",
    "/api/device-pairing/grants",
    "/api/device-pairing/exchange",
];

/// POST endpoints that inspect or verify without mutating server state.
const READ_ONLY_POST_EXACT: &[&str] = &[
    "/api/network/verify-handshake",
    "/api/users/verify",
    "/api/settings/video-info",
    "/api/settings/audio-gain",
];

fn is_device_pairing_delivery(path: &str) -> bool {
    path.starts_with("/api/device-pairing/desktop-sessions/")
        && (path.ends_with("/deliver") || path.ends_with("/verify"))
}

/// Test whether an IPv4 address falls in the RFC 6598 carrier-grade NAT range
/// (100.64.0.0/10), which Tailscale uses for its internal "100.x.x.x" addresses.
/// std's `is_private()` does not include this range.
fn is_cgnat(v4: &std::net::Ipv4Addr) -> bool {
    let octets = v4.octets();
    octets[0] == 100 && (octets[1] & 0b1100_0000) == 64
}

/// Test whether an IPv6 address is in Tailscale's ULA range (fd7a:115c:a1e0::/48).
fn is_tailscale_v6(v6: &std::net::Ipv6Addr) -> bool {
    let segs = v6.segments();
    segs[0] == 0xfd7a && segs[1] == 0x115c && segs[2] == 0xa1e0
}

/// Classify an IP address into an access level.
pub fn classify_ip(ip: &std::net::IpAddr) -> &'static str {
    match ip {
        std::net::IpAddr::V4(v4) => {
            if v4.is_loopback() {
                "localhost"
            } else if v4.is_private() || v4.is_link_local() || is_cgnat(v4) {
                "local_network"
            } else {
                "public"
            }
        }
        std::net::IpAddr::V6(v6) => {
            if v6.is_loopback() {
                "localhost"
            } else if is_tailscale_v6(v6) {
                "local_network"
            } else {
                // Check for IPv4-mapped IPv6 (::ffff:127.0.0.1, etc.)
                if let Some(v4) = v6.to_ipv4_mapped() {
                    if v4.is_loopback() {
                        return "localhost";
                    }
                    if v4.is_private() || v4.is_link_local() || is_cgnat(&v4) {
                        return "local_network";
                    }
                }
                "public"
            }
        }
    }
}

/// Read the `network.allow_settings_local_network` opt-in flag from `settings.json`.
///
/// When `true`, the owner has explicitly allowed local-network devices (with a valid
/// JWT) to reach the otherwise localhost-only settings/network/user endpoints. Returns
/// `false` (the secure default) if the file or key is missing or unreadable, so a
/// missing/corrupt settings file fails closed.
fn lan_settings_opt_in(data_dir: &Path) -> bool {
    let path = data_dir.join("settings.json");
    std::fs::read_to_string(&path)
        .ok()
        .and_then(|s| serde_json::from_str::<serde_json::Value>(&s).ok())
        .and_then(|v| {
            v.get("network")?
                .get("allow_settings_local_network")?
                .as_bool()
        })
        .unwrap_or(false)
}

// ─── Access tier ────────────────────────────────────────────────────────────

/// Typed access tier derived from a client IP address.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AccessTier {
    Localhost,
    LocalNetwork,
    Public,
}

impl AccessTier {
    /// Classify a client IP into an access tier.
    pub fn from_ip(ip: &std::net::IpAddr) -> Self {
        match classify_ip(ip) {
            "localhost" => Self::Localhost,
            "local_network" => Self::LocalNetwork,
            _ => Self::Public,
        }
    }
}

// ─── Layer ──────────────────────────────────────────────────────────────────

#[derive(Clone)]
pub struct AccessControlLayer {
    pub jwt_secret: String,
    /// Data directory, used to read the `allow_settings_local_network` opt-in flag.
    pub data_dir: PathBuf,
    pub db: DbPool,
}

impl<S> Layer<S> for AccessControlLayer {
    type Service = AccessControlService<S>;

    fn layer(&self, inner: S) -> Self::Service {
        AccessControlService {
            inner,
            jwt_secret: self.jwt_secret.clone(),
            data_dir: self.data_dir.clone(),
            db: self.db.clone(),
        }
    }
}

// ─── Service ────────────────────────────────────────────────────────────────

#[derive(Clone)]
pub struct AccessControlService<S> {
    inner: S,
    jwt_secret: String,
    data_dir: PathBuf,
    db: DbPool,
}

impl<S> Service<Request<Body>> for AccessControlService<S>
where
    S: Service<Request<Body>, Response = Response<Body>> + Clone + Send + 'static,
    S::Future: Send + 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, req: Request<Body>) -> Self::Future {
        let mut inner = self.inner.clone();
        // Swap so the clone is the "not ready" one
        std::mem::swap(&mut self.inner, &mut inner);

        let jwt_secret = self.jwt_secret.clone();
        let data_dir = self.data_dir.clone();
        let db = self.db.clone();

        Box::pin(async move {
            let path = req.uri().path().to_string();
            let method = req.method().as_str().to_string();

            // Always allow OPTIONS (CORS preflight)
            if method == "OPTIONS" {
                return inner.call(req).await;
            }

            // Skip access control for exempt endpoints (prefix match)
            for prefix in EXEMPT_PREFIXES {
                if path.starts_with(prefix) {
                    return inner.call(req).await;
                }
            }

            // Skip access control for exempt endpoints (exact match)
            if EXEMPT_EXACT.contains(&path.as_str()) {
                return inner.call(req).await;
            }

            // Extract client IP from ConnectInfo extension
            let client_ip = req
                .extensions()
                .get::<ConnectInfo<SocketAddr>>()
                .map(|ci| ci.0.ip())
                .unwrap_or(std::net::IpAddr::V4(std::net::Ipv4Addr::UNSPECIFIED));

            let access_level = classify_ip(&client_ip);

            // Localhost has full access
            if access_level == "localhost" {
                return inner.call(req).await;
            }

            // Decide whether a media-scoped `?token=` grants access to THIS request.
            // Full session tokens are never accepted from URLs. A media token is
            // read-only and only honored for GET requests on `/api/images/...` —
            // the one media path that requires a token (streams, cast-media, share,
            // /thumbnails and /watch are already exempt). This lets the frontend put
            // a short-lived media token in <img>/<video> URLs instead of the 30-day
            // session JWT, without breaking any media that needs auth.
            let query_token_grants = |token: &str| -> bool {
                match decode_jwt(token, &jwt_secret) {
                    Ok(claims) if claims.is_media_scoped() => {
                        method == "GET"
                            && path.starts_with("/api/images/")
                            && refresh_claim_identity(claims, &db)
                                .is_some_and(|claims| claim_allows_source(&claims, access_level))
                    }
                    Ok(_) => false,
                    Err(_) => false,
                }
            };

            // Bearer header: accept full session tokens only. A media token is
            // rejected here so it can never authenticate a real API call.
            let bearer_claims = req
                .headers()
                .get("authorization")
                .and_then(|v| v.to_str().ok())
                .and_then(|auth| {
                    auth.strip_prefix("Bearer ")
                        .or_else(|| auth.strip_prefix("bearer "))
                })
                .and_then(|token| decode_jwt(token, &jwt_secret).ok())
                .filter(|claims| claims.scope.is_none())
                .and_then(|claims| refresh_claim_identity(claims, &db))
                .filter(|claims| claim_allows_source(claims, access_level));

            // Query parameters are accepted only for short-lived media tokens.
            let has_valid_query_jwt = req
                .uri()
                .query()
                .and_then(|q| q.split('&').find_map(|pair| pair.strip_prefix("token=")))
                .map(query_token_grants)
                .unwrap_or(false);
            let has_valid_jwt = bearer_claims.is_some() || has_valid_query_jwt;

            // Paths explicitly allowed from the network even though they sit under a
            // localhost-only prefix: pairing handshake, login/verify, and user-preference
            // settings the LAN/Android client legitimately needs.
            let is_exempt_path = LOCALHOST_EXEMPTIONS
                .iter()
                .any(|exempt| path == *exempt || path.starts_with(&format!("{}/", exempt)))
                || (method == "POST" && is_device_pairing_delivery(&path));

            // Localhost-only endpoints (settings / network / user management) stay
            // restricted even WITH a valid JWT. Localhost already returned above, so
            // reaching here means a non-localhost caller. A paired device's token must
            // NOT be able to change network exposure, rewrite global settings, or manage
            // users — UNLESS the owner has explicitly opted in from the host machine via
            // `network.allow_settings_local_network`, and only from the local network
            // (never from the public internet).
            let is_localhost_only = LOCALHOST_ONLY_PREFIXES
                .iter()
                .any(|prefix| path.starts_with(prefix))
                && !is_exempt_path;

            let is_mutating = !matches!(method.as_str(), "GET" | "HEAD")
                && !(method == "POST" && READ_ONLY_POST_EXACT.contains(&path.as_str()));
            if is_mutating
                && bearer_claims
                    .as_ref()
                    .is_some_and(|claims| !claims.can_write)
            {
                let response = (
                    StatusCode::FORBIDDEN,
                    axum::Json(serde_json::json!({
                        "error": "Write access required",
                        "detail": "This credential is read-only."
                    })),
                )
                    .into_response();
                return Ok(response);
            }

            if is_localhost_only {
                let opt_in_allowed = !path.starts_with("/api/direct-files")
                    && !path.starts_with("/api/device-pairing/desktop-sessions")
                    && !path.starts_with("/api/device-pairing/devices")
                    && has_valid_jwt
                    && access_level == "local_network"
                    && lan_settings_opt_in(&data_dir);

                if !opt_in_allowed {
                    let response = (
                        StatusCode::FORBIDDEN,
                        axum::Json(serde_json::json!({
                            "error": "This endpoint is only accessible from localhost",
                            "detail": "Settings, network, and user management are restricted to the host machine. Enable 'Allow settings changes over local network' on the host to manage them from a LAN device."
                        })),
                    )
                        .into_response();
                    return Ok(response);
                }

                // Opted in: a local-network device with a valid JWT may proceed.
                return inner.call(req).await;
            }

            // Authenticated requests (valid JWT) may access everything that is not
            // localhost-only (handled above): images, tags, media, casting, user prefs, etc.
            if has_valid_jwt {
                return inner.call(req).await;
            }

            // No valid JWT and not localhost → block with 401.
            // Localhost-exempted paths (e.g. verify-handshake, login) pass through so a
            // client can obtain a token in the first place.
            if !is_exempt_path {
                let response = (
                    StatusCode::UNAUTHORIZED,
                    axum::Json(serde_json::json!({
                        "error": "Authentication required",
                        "detail": "Non-localhost requests require a valid JWT token. Pair via QR code to obtain one."
                    })),
                )
                    .into_response();
                return Ok(response);
            }

            inner.call(req).await
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::server::middleware::auth::{create_jwt, create_media_jwt};
    use tower::{service_fn, ServiceExt};

    #[tokio::test]
    async fn paired_lan_client_can_stop_retired_optical_flow_playback() {
        let secret = "optical-flow-stop-test-secret";
        let token = create_jwt(0, "paired-device", "local_network", true, secret).unwrap();
        let data_dir = std::env::temp_dir().join(format!(
            "localbooru-access-control-test-{}",
            uuid::Uuid::new_v4()
        ));
        std::fs::create_dir_all(&data_dir).unwrap();
        let db = crate::db::pool::create_main_pool(&data_dir).unwrap();
        crate::db::schema::init_main_db(&db.get().unwrap()).unwrap();

        let inner = service_fn(|_request: Request<Body>| async move {
            Ok::<_, std::convert::Infallible>(Response::new(Body::empty()))
        });
        let service = AccessControlLayer {
            jwt_secret: secret.into(),
            data_dir: data_dir.clone(),
            db: db.clone(),
        }
        .layer(inner);

        let mut request = Request::builder()
            .method("POST")
            .uri("/api/settings/optical-flow/stop")
            .header("authorization", format!("Bearer {token}"))
            .body(Body::empty())
            .unwrap();
        request
            .extensions_mut()
            .insert(ConnectInfo(SocketAddr::from(([192, 168, 1, 25], 50000))));

        let response = service.oneshot(request).await.unwrap();
        assert_eq!(response.status(), StatusCode::OK);
        let _ = std::fs::remove_dir_all(data_dir);
    }

    #[tokio::test]
    async fn direct_file_capabilities_are_never_available_to_lan_clients() {
        let secret = "direct-file-localhost-test-secret";
        let token = create_jwt(0, "paired-device", "local_network", true, secret).unwrap();
        let data_dir = std::env::temp_dir().join(format!(
            "localbooru-direct-file-access-test-{}",
            uuid::Uuid::new_v4()
        ));
        std::fs::create_dir_all(&data_dir).unwrap();
        let db = crate::db::pool::create_main_pool(&data_dir).unwrap();
        crate::db::schema::init_main_db(&db.get().unwrap()).unwrap();
        std::fs::write(
            data_dir.join("settings.json"),
            r#"{"network":{"allow_settings_local_network":true}}"#,
        )
        .unwrap();

        let inner = service_fn(|_request: Request<Body>| async move {
            Ok::<_, std::convert::Infallible>(Response::new(Body::empty()))
        });
        let service = AccessControlLayer {
            jwt_secret: secret.into(),
            data_dir: data_dir.clone(),
            db: db.clone(),
        }
        .layer(inner);
        let mut request = Request::builder()
            .method("GET")
            .uri("/api/direct-files/unguessable-token")
            .header("authorization", format!("Bearer {token}"))
            .body(Body::empty())
            .unwrap();
        request
            .extensions_mut()
            .insert(ConnectInfo(SocketAddr::from(([192, 168, 1, 25], 50000))));

        let response = service.oneshot(request).await.unwrap();
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        let _ = std::fs::remove_dir_all(data_dir);
    }

    #[tokio::test]
    async fn lan_clients_cannot_mint_phone_pairing_offers() {
        let data_dir = std::env::temp_dir().join(format!(
            "localbooru-qr-offer-access-test-{}",
            uuid::Uuid::new_v4()
        ));
        std::fs::create_dir_all(&data_dir).unwrap();
        let db = crate::db::pool::create_main_pool(&data_dir).unwrap();
        crate::db::schema::init_main_db(&db.get().unwrap()).unwrap();
        let inner = service_fn(|_request: Request<Body>| async move {
            Ok::<_, std::convert::Infallible>(Response::new(Body::empty()))
        });
        let service = AccessControlLayer {
            jwt_secret: "qr-offer-test-secret".into(),
            data_dir: data_dir.clone(),
            db,
        }
        .layer(inner);
        let mut request = Request::builder()
            .method("GET")
            .uri("/api/network/qr-data")
            .body(Body::empty())
            .unwrap();
        request
            .extensions_mut()
            .insert(ConnectInfo(SocketAddr::from(([192, 168, 1, 25], 50000))));

        let response = service.oneshot(request).await.unwrap();
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        let _ = std::fs::remove_dir_all(data_dir);
    }

    #[tokio::test]
    async fn full_query_tokens_and_read_only_mutations_are_rejected() {
        let secret = "query-token-and-write-test-secret";
        let data_dir = std::env::temp_dir().join(format!(
            "localbooru-query-token-access-test-{}",
            uuid::Uuid::new_v4()
        ));
        std::fs::create_dir_all(&data_dir).unwrap();
        let db = crate::db::pool::create_main_pool(&data_dir).unwrap();
        crate::db::schema::init_main_db(&db.get().unwrap()).unwrap();
        db.get()
            .unwrap()
            .execute(
                "INSERT INTO users (id, username, password_hash, is_active, access_level, can_write) VALUES (5, 'reader', 'unused', 1, 'local_network', 0)",
                [],
            )
            .unwrap();
        let full_token = create_jwt(5, "reader", "local_network", false, secret).unwrap();
        let media_token = create_media_jwt(5, "reader", "local_network", None, secret).unwrap();
        let inner = service_fn(|_request: Request<Body>| async move {
            Ok::<_, std::convert::Infallible>(Response::new(Body::empty()))
        });
        let service = AccessControlLayer {
            jwt_secret: secret.into(),
            data_dir: data_dir.clone(),
            db,
        }
        .layer(inner);
        let mut full_query = Request::builder()
            .method("GET")
            .uri(format!("/api/images/1?token={full_token}"))
            .body(Body::empty())
            .unwrap();
        full_query
            .extensions_mut()
            .insert(ConnectInfo(SocketAddr::from(([192, 168, 1, 25], 50000))));
        assert_eq!(
            service.clone().oneshot(full_query).await.unwrap().status(),
            StatusCode::UNAUTHORIZED
        );

        let mut media_query = Request::builder()
            .method("GET")
            .uri(format!("/api/images/1?token={media_token}"))
            .body(Body::empty())
            .unwrap();
        media_query
            .extensions_mut()
            .insert(ConnectInfo(SocketAddr::from(([192, 168, 1, 25], 50000))));
        assert_eq!(
            service.clone().oneshot(media_query).await.unwrap().status(),
            StatusCode::OK
        );

        let mut read_only_post = Request::builder()
            .method("POST")
            .uri("/api/users/verify")
            .header("authorization", format!("Bearer {full_token}"))
            .body(Body::empty())
            .unwrap();
        read_only_post
            .extensions_mut()
            .insert(ConnectInfo(SocketAddr::from(([192, 168, 1, 25], 50000))));
        assert_eq!(
            service
                .clone()
                .oneshot(read_only_post)
                .await
                .unwrap()
                .status(),
            StatusCode::OK
        );

        let mut mutation = Request::builder()
            .method("POST")
            .uri("/api/images/1/favorite")
            .header("authorization", format!("Bearer {full_token}"))
            .body(Body::empty())
            .unwrap();
        mutation
            .extensions_mut()
            .insert(ConnectInfo(SocketAddr::from(([192, 168, 1, 25], 50000))));
        assert_eq!(
            service.oneshot(mutation).await.unwrap().status(),
            StatusCode::FORBIDDEN
        );
        let _ = std::fs::remove_dir_all(data_dir);
    }
}
